nemo/collections/asr/models/enhancement_models.py

```diff
@@ -0,0 +1,440 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+import os
+import tempfile
+from typing import Dict, List, Optional, Union
+
+import librosa
+import soundfile as sf
+import torch
+from omegaconf import DictConfig
+from pytorch_lightning import Trainer
+from tqdm import tqdm
+
+from nemo.collections.asr.data import audio_to_audio_dataset, audio_to_text_dataset
+from nemo.collections.asr.models.audio_processing_model import AudioProcessingModel
+from nemo.collections.asr.parts.utils.audio_utils import ChannelSelectorType
+from nemo.core.classes.common import PretrainedModelInfo, typecheck
+from nemo.core.neural_types import AudioSignal, LengthsType, NeuralType
+from nemo.utils import logging
+from nemo.utils.decorators import experimental
+
+__all__ = ['EncMaskDecAudioProcessingModel']
+
+
+@experimental
+class EncMaskDecAudioProcessingModel(AudioProcessingModel):
+    """Base class for encoder-mask-decoder audio processing models."""
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None):
+        # Get global rank and total number of GPU workers for IterableDataset partitioning, if applicable
+        # Global_rank and local_rank is set by LightningModule in Lightning 1.2.0
+        self.world_size = 1
+        if trainer is not None:
+            self.world_size = trainer.world_size
+
+        super().__init__(cfg=cfg, trainer=trainer)
+        self.sample_rate = self._cfg.sample_rate
+        self.num_outputs = self._cfg.num_outputs
+        self.encoder = EncMaskDecAudioProcessingModel.from_config_dict(self._cfg.encoder)
+        self.mask = EncMaskDecAudioProcessingModel.from_config_dict(self._cfg.mask)
+        self.processor = EncMaskDecAudioProcessingModel.from_config_dict(self._cfg.processor)
+        self.decoder = EncMaskDecAudioProcessingModel.from_config_dict(self._cfg.decoder)
+        self.loss = EncMaskDecAudioProcessingModel.from_config_dict(self._cfg.loss)
```

Comment:
i can imagine these class of problems having multiple losses. It might be beneficial to support that .. some thing like.. https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/models/ssl_models.py#L85 ?